transport.go

```diff
@@ -22,29 +21,6 @@ var Protocol = ma.Protocol{
 
 func init() {
 	ma.AddProtocol(Protocol)
-
-	// Add dialer transport
-	const unspecific = "/p2p-circuit/ipfs"
-	const proto = "/ipfs/p2p-circuit/ipfs"
-
-	tps := addrutil.SupportedTransportStrings
-
-	err := addrutil.AddTransport(unspecific)
-	if err != nil {
-		panic(err)
-	}
-
-	err = addrutil.AddTransport(proto)
-	if err != nil {
-		panic(err)
-	}
-
-	for _, tp := range tps {
```

Comment:
is there equivalent code for this or is it handled in a different manner now?

Reply:
It's handled by the swarm internally. See: https://github.com/libp2p/go-addr-util/pull/11

Reply:
is the swarm explicitly aware of relay addresses or does it ask the transports dynamically to respond whether they support the address?
If so, how are encapsulations handled?

Reply:
The swarm asks its transports.

For listening, we just figure out which transport is responsible using the algorithm described [here](https://github.com/libp2p/go-libp2p-transport/blob/88e91d2b047d46b35012678dfa23b711d2ba5f8f/transport.go#L81-L94) and ask it to listen on the address.

For dialing, we lookup the appropriate transport (same algorithm) and call the `CanDial` function on it ([code](https://github.com/libp2p/go-libp2p-swarm/blob/feat/refactor/swarm_dial.go#L256)). We could just try dialing but this way allows us to filter out addresses we definitely can't dial higher up the stack.

Reply:
Not clear it will have the same behaviour -- for example we want to pick up relay for `/ip4/X.X.X.X/tcp/YY/p2p-circuit/ip4/X.X.X.X/tcp/YY/ipfs/Qmfoo` (that's explicit relay dialing).
What will the algorithm pick here?

Reply:
I don't see how that's related to *this* function but it will pick the relay transport. It'll:

1. Iterate over the protocols (ip4, tcp, ipfs, p2p-circuit, ip4, tcp, ipfs).
2. When it gets to the p2p-circuit protocol, it will notice that the associated transport is a `Proxy` transport and will pass the address off to that transport.

Reply:
ok, being a proxy is the key feature.